storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -800,8 +802,23 @@ private Void deleteTypeIfExists(StorageAction<Boolean, IOException> delete, Stri
         try {
             if (delete.execute())
                 LOGGER.info("Deleted {} {}.", fileType, file.getAbsolutePath());
-            else if (logIfMissing)
-                LOGGER.info("Failed to delete {} {} because it does not exist.", fileType, file.getAbsolutePath());
+            else {
+                if (logIfMissing) {
+                    LOGGER.info("Failed to delete {} {} because it does not exist.", fileType, file.getAbsolutePath());
+                }
+
+                // During alter log dir, the log segment may be moved to a new directory, so async delete may fail.
+                // Fallback to delete the file in the new directory to avoid orphan file.
+                Pattern dirPattern = Pattern.compile("^(\\S+)-(\\S+)\\.(\\S+)-(delete|future)");
```

Comment:
1. Why does it contain `delete` in the end?
2. Unfortunately, the topic name could contain `-` or `.`, so it's unsafe to do regex like this.

I'm thinking we can pass `topicPartition` as parameter into `deleteTypeIfExists` so that we don't have to do further regex like this. And just verify if fileName.endsWith("future") because the normal folder name should always ends with a number (partition number), instead of "future". WDYT?

Reply:
1. Yeah, I misunderstood file name and folder name. Removed `-delete` folder case.
2. We can use `topicPartition`, but we use similar regular expression in other place. Do we also need to update it?

https://github.com/apache/kafka/blob/269b457d30940e51f532d6a1b616b9506f87232f/core/src/main/scala/kafka/log/LocalLog.scala#L586-L588

Reply:
> We can use topicPartition, but we use similar regular expression in other place. Do we also need to update it?

Thanks for pointing out these to me. I was wrong. I tried to match with topic name containing `.` or `-`, they both match correctly. That means we can use this regex directly.